### DISCUSSION TITLE
Using dependencies from the project within Behat listener

### DISCUSSION THREAD

**Discussion by monitaurus** [Behat/Behat#1579] [Q&A]
Hi there, I need to run some test preparation before an after every scenario in my tests.
I've got like 10 suites, one per domain, and I didn't find a way to load a specific context by default in all those suites.
Currently I load a context in each of those suites, but I try to find something _"prettier"_.

I found out recently that I can create a listener that could subscribe to before and after scenarios events.
I took this listner [SaveFailingPagesListener](https://github.com/ingenerator/behat-support/blob/1.x/src/Extension/SaveFailingPagesExtension/SaveFailingPagesListener.php) from this extension [SaveFailingPagesExtension](https://github.com/ingenerator/behat-support/blob/66028a7f3ed4195a3c042bc8e49966a197753144/src/Extension/SaveFailingPagesExtension.php#L49-L58) as example.

I try to create one, but my issue is that I need dependencies from my project (`src/`) within my tests.
From what I understand, Behat's run in it's own autoload / container realm, so the example above, the definition will only be able to instanciate with dependencies from Behat's container.

Is there a way to use dependencies from the project within Behat listener?
And if not, is there a way to load by default a context?

**Comment by acoulton** [comment]
@monitaurus I'm not sure I've 100% understood as you mention both autoload and container.

### Autoloading
Behat has a custom autoloader for Context classes, but it also registers the standard composer autoloader - so if it's just that you need to autoload classes under `src/` that should work automatically if they are autoloadable by composer.

### Service container

If it's not enough to just autoload the classes and you need to access instances of those classes / services from your runtime service container (e.g. configured the same way they would be in your app runtime code) then you are correct that can't be done directly from Behat.

You have probably got two options:

* Duplicate the service configurations in the behat container (probably not ideal, although might be worth it if they need some different configuration).
* Provide a way to access the runtime service container e.g. as a global singleton and have your extension inject *that* container into the contexts. Exactly how you do that will depend on the DI container you are using but you could look at our [KohanaDependenciesExtension](https://github.com/ingenerator/behat-support/blob/66028a7f3ed4195a3c042bc8e49966a197753144/src/Extension/KohanaDependenciesExtension.php) and the [DependenciesAwareContext interface](https://github.com/ingenerator/behat-support/blob/66028a7f3ed4195a3c042bc8e49966a197753144/src/Extension/KohanaDependenciesExtension/DependenciesAwareContext.php) and [KohanaDependenciesInitialiser](https://github.com/ingenerator/behat-support/blob/66028a7f3ed4195a3c042bc8e49966a197753144/src/Extension/KohanaDependenciesExtension/KohanaDependenciesInitialiser.php) for an example where the app container is available from `Dependencies::instance()` and then passed into contexts that implement a `setDependencies(DependencyContainer $dependencies)` method. That method is called just after the context is initialised and is responsible for assigning the specific dependencies that context needs (e.g. with `$container->get(...)`).

If you are using symfony, then the https://github.com/FriendsOfBehat/SymfonyExtension may also be worth looking at - I believe it allows you to define your contexts as symfony services (and presumably therefore bind them to other services) but I have not used it.

### Default contexts for different suites

Not sure here if you mean `context` as in a Behat `Context` class, or as a more generic "runtime context of services and configuration". If you want to load a different Behat `Context` class for each suite, the docs for configuring that are [here](https://docs.behat.org/en/latest/user_guide/context.html#multiple-contexts)
